include/cuda/std/memory_resource

```diff
@@ -0,0 +1,268 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_MEMORY_RESOURCE
+#define _CUDA_MEMORY_RESOURCE
+
+#include "cstddef"
+#include "detail/__config"
+#include "detail/__pragma_push"
+#include "stream"
+#include "version"
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+/**
+ * @brief Specifies the kind of memory of an allocation.
+ *
+ * Memory allocation kind determines where memory can be accessed and the
+ * performance characteristics of accesses.
+ *
+ */
+enum class memory_kind {
+  device,  ///< Device memory accessible only from device
+  unified, ///< Unified memory accessible from both host and device
+  pinned,  ///< Page-locked system memory accessible from both host and device
+  host     ///< System memory only accessible from host code
+};
+
+/**
+ * @brief Abstract interface for synchronous memory allocation.
+ *
+ * @tparam _Kind The `memory_kind` of the allocated memory.
+ */
+template <memory_kind _Kind> class memory_resource {
+public:
+  static constexpr memory_kind memory_kind = _Kind;
+
+  static constexpr std::size_t default_alignment =
+      (_Kind == memory_kind::host) ? alignof(std::max_align_t) : 256;
+
+  virtual ~memory_resource() = default;
+
+  /**
+   * @brief Allocates storage of size at least `__bytes` bytes.
+   *
+   * The returned storage is aligned to the specified `__alignment` if such
+   * alignment is supported, and to `default_alignment` otherwise.
+   *
+   * @throws If storage of the requested size and alignment cannot be obtained.
+   *
+   * @param __bytes The size in bytes of the allocation
+   * @param __alignment The alignment of the allocation
+   * @return Pointer to the requested storage
+   */
+  void *allocate(std::size_t __bytes,
+                 std::size_t __alignment = default_alignment) {
+    return do_allocate(__bytes, __alignment);
+  }
+
```

Comment:
```
/// Value you can't get an alignment lower than, but won't throw
virtual std::size_t guaranteed_alignment() = 0;

/// Alignment value you can't go higher than, otherwise throw
virtual std::size_t maximum_alignment() = 0;
```

Reply:
constexpr?